### PR TITLE
#753 Super critical exception workaround.

### PR DESCRIPTION
**Pull request by ArtemBuryko**
Wrapped into the try-attempt loop with 20 sec delay between

### CODE REVIEW

File: Indexing.Core/Indexer.cs [hunk 1]
```diff
@@ -40,19 +43,26 @@ public Indexer(IWeb3Tracer web3Tracer, ILogger logger, string dbConnectionString
...
+                catch (Exception ex)
+                {
+                    Logger.LogCritical($"Super critical exception occured. Exiting the app. Ex: {ex.Message}");
 
-                throw ex;
+                    if (attempts >= MaxAttempts)
+                        throw ex;
+                    else
+                        System.Threading.Thread.Sleep(DelayAttempts * 1000);
+                }
```
Comment (borseno) [inline Indexing.Core/Indexer.cs:62]: maybe use Task.Delay to avoid throttling threads?

File: Indexing.Core/Indexer.cs [hunk 2]
```diff
@@ -80,24 +90,34 @@ public async Task CommitIndexedBlocks()
...
+                        Logger.LogInformation("Successfully saved");
+                    }
+                    catch (Exception ex)
+                    {
+                        Logger.LogCritical("Error while saving BlockQueue into database. ex: " + ex.Message);
+
+                        if (attempts >= MaxAttempts)
+                            Environment.Exit(1);
+                        else
+                            System.Threading.Thread.Sleep(DelayAttempts * 1000);
```
Comment (borseno) [inline Indexing.Core/Indexer.cs:114]: same comment as above, we should use System.Task.Delay vs Thread.Sleep in async functions. or explain why this is used here?

File: Indexing.Core/Indexer.cs [hunk 3]
```diff
@@ -126,80 +146,101 @@ public async Task StartMonitorNewBlocks(double milliSecondsToSleepBeforeIteratio
...
+                        await CommitIndexedBlocks();
+                }
+                catch (Exception ex) // if exception wasn`t handled in IndexBlock method - this is a Rpc call Exception
+                {
+                    Logger.LogCritical($"Super critical exception occured. Exiting the app. Ex: {ex.Message}");
+                    
+                    if (attempts >= MaxAttempts)
+                        Environment.Exit(1);
+                    else
+                        System.Threading.Thread.Sleep(DelayAttempts * 1000);
```
Comment (borseno) [inline Indexing.Core/Indexer.cs:186]: same comment as above, we should use Task.Delay vs Thread.Sleep in async functions.

File: Indexing.Core/Indexer.cs [hunk 4]
```diff
@@ -126,80 +146,101 @@ public async Task StartMonitorNewBlocks(double milliSecondsToSleepBeforeIteratio
...
+                        if (!BlockQueue.Any(v => (ulong)v.BlockNumber == i))
+                            tasks.Add(IndexBlock(i));
+                    }
+
+                    await Task.WhenAll(tasks);
+
+                    if (BlockQueue.Count >= QueueSize)
+                        await CommitIndexedBlocks();
+                }
+                catch (Exception ex) // if exception wasn`t handled in IndexBlock method - this is a Rpc call Exception
```
Comment (borseno) [inline Indexing.Core/Indexer.cs:179]: It seems that this code is repeating, could you extract repeating code into a common function to keep the code DRY.

File: Indexing.Core/Indexer.cs [hunk 5]
```diff
@@ -126,80 +146,101 @@ public async Task StartMonitorNewBlocks(double milliSecondsToSleepBeforeIteratio
...
+                    Logger.LogInformation($"Block {blockNubmer} added to InsertBlockQueue");
+                }
+                catch (Exception ex)
+                {
+                    Logger.LogError($"GetBlockTransactions() Failed on block {blockNubmer}. Ex: {ex.Message}");
+                    
+                    if (attempts >= MaxAttempts)
+                        throw ex;
+                    else
+                        System.Threading.Thread.Sleep(DelayAttempts * 1000);
```
Comment (borseno) [inline Indexing.Core/Indexer.cs:240]: Same comment as above - please look at all the changes and make sure they comply with the comments made earlier

**Review by borseno — CHANGES_REQUESTED** [review]

**Review by borseno — CHANGES_REQUESTED** [review]

File: Indexing.Core/Indexer.cs [hunk 6]
```diff
@@ -351,6 +335,14 @@ private string GetMethodIdFromInput(string value)
             return value.Substring(0, value.Length > 10 ? 10 : value.Length);
         }
 
+        private async void ThrowOrDelay(int attempts, Exception ex)
```
[the remark this answers is not in the document]
Comment (ArtemBuryko) [reply Indexing.Core/Indexer.cs]: fixed

**Pull request merged**